File: packages/entity-sync-core/src/messages.ts
/**
 * PSyncV1 message types, mirroring `pes-protocol/src/messages.rs`.
 *
 * Rust's `serde` derive uses the externally-tagged enum representation by
 * default:
 * - A variant with fields encodes as a single-key map: `{ VariantName: { ...fields } }`.
 * - A unit variant (no fields, e.g. `ClientMessage::Ping`, `Op::Delete`)
 *   encodes as a bare string: `"VariantName"`, NOT `{ VariantName: null }`.
 *
 * Newtype wrappers like `BucketId(pub String)` and `PgLsn(pub u64)`
 * serialize transparently as their inner value — a `BucketId` is just a
 * JSON/MessagePack string on the wire, never `{ "0": "..." }`.
 *
 * These shapes were verified empirically against the real Rust encoder
 * (`rmp_serde::to_vec_named`), not inferred from the Rust source alone —
 * see the `wire_shape_check` scratch test used during development.
 */

export const PROTOCOL_VERSION = 1;
export const ERROR_CODE_UNSUPPORTED_PROTOCOL_VERSION = 4000;

/** A single op appended to a bucket op log. Mirrors `pes_core::BucketOp`. */
export interface BucketOp {
  lsn: number;
  bucket_id: string;
  entity_type: string;
  entity_id: string;
  op: Op;
}

/**
 * Mirrors `pes_core::Op`. `Delete` is a unit variant → bare string `"Delete"`.
 * `Upsert`/`CrdtPatch` carry a payload → single-key object.
 */
export type Op =
  | { Upsert: unknown }
  | "Delete"
  | { CrdtPatch: Uint8Array };

export type ServerMessage =
  | {
      SnapshotBegin: {
        bucket_id: string;
        total_rows: number;
        protocol_version: number;
      };
    }
  | {
      SnapshotBatch: {
        bucket_id: string;
        rows: unknown[];
        offset: number;
      };
    }
  | {
      SnapshotComplete: {
        bucket_id: string;
        checksum: number;
      };
    }
  | {
      Delta: {
        bucket_id: string;
        ops: BucketOp[];
        lsn: number;
      };
    }
  | {
      Checkpoint: {
        lsn: number;
        bucket_checksums: Record<string, number>;
      };
    }
  | {
      Keepalive: {
        server_time_ms: number;
      };
    }
  | {
      Error: {
        code: number;
        message: string;
      };
    };

export type ClientMessage =
  | {
      Subscribe: {
        buckets: string[];
        token: string;
        resume_lsn: number | null;
        protocol_version: number;
      };
    }
  | {
      Ack: {
        lsn: number;
      };
    }
  | {
      Write: {
        entity_type: string;
        entity_id: string;
        op: Op;
      };
    }
  /** `Ping` is a unit variant → bare string `"Ping"`. */
  | "Ping";

File: crates/pes-protocol/src/messages.rs
//! PSyncV1 message types: [`ServerMessage`] and [`ClientMessage`].

use std::collections::HashMap;

use pes_core::{BucketChecksum, BucketId, BucketOp, Op, PgLsn};
use serde::{Deserialize, Serialize};

/// Current PSyncV1 protocol version. Sent in [`ClientMessage::Subscribe`]
/// and echoed back in [`ServerMessage::SnapshotBegin`].
pub const PROTOCOL_VERSION: u8 = 1;

/// Error code for [`ServerMessage::Error`] when a client's
/// `protocol_version` is not supported by this server.
pub const ERROR_CODE_UNSUPPORTED_PROTOCOL_VERSION: u16 = 4000;

/// Messages sent from the sync server to a client.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[non_exhaustive]
pub enum ServerMessage {
    /// Sent once per bucket at the start of initial snapshot delivery.
    SnapshotBegin {
        /// The bucket this snapshot is for.
        bucket_id: BucketId,
        /// Total number of rows the client should expect across all
        /// [`ServerMessage::SnapshotBatch`] messages for this bucket.
        total_rows: u64,
        /// The protocol version this server is using for this connection.
        protocol_version: u8,
    },
    /// One page of a bucket's snapshot.
    SnapshotBatch {
        /// The bucket this batch belongs to.
        bucket_id: BucketId,
        /// The rows in this batch, as JSON objects.
        rows: Vec<serde_json::Value>,
        /// 0-based offset of this batch within the bucket's snapshot.
        offset: u64,
    },
    /// Sent once per bucket after the last [`ServerMessage::SnapshotBatch`],
    /// carrying the checksum the client should verify against.
    SnapshotComplete {
        /// The bucket whose snapshot just completed.
        bucket_id: BucketId,
        /// The bucket's checksum, for the client to verify integrity.
        checksum: BucketChecksum,
    },
    /// A batch of live ops for one bucket, delivered after the initial
    /// snapshot phase.
    Delta {
        /// The bucket these ops belong to.
        bucket_id: BucketId,
        /// The ops themselves, in order.
        ops: Vec<BucketOp>,
        /// The LSN of the last op in `ops` — the client should `Ack` this.
        lsn: PgLsn,
    },
    /// Periodic checkpoint summarizing every subscribed bucket's current
    /// state, for client-side integrity verification without needing a
    /// fresh snapshot.
    Checkpoint {
        /// The LSN this checkpoint was taken at.
        lsn: PgLsn,
        /// Every subscribed bucket's checksum as of `lsn`.
        bucket_checksums: HashMap<BucketId, BucketChecksum>,
    },
    /// Periodic keepalive, also usable by the client to measure clock skew.
    Keepalive {
        /// Server wall-clock time, milliseconds since the Unix epoch.
        server_time_ms: u64,
    },
    /// A fatal or informational error. See `ERROR_CODE_*` constants for
    /// well-known codes.
    Error {
        /// A numeric error code (see `ERROR_CODE_*` constants).
        code: u16,
        /// A human-readable description.
        message: String,
    },
}

/// Messages sent from a client to the sync server.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[non_exhaustive]
pub enum ClientMessage {
    /// Sent once at connection start to request buckets and authenticate.
    Subscribe {
        /// The bucket ids the client wants to subscribe to.
        buckets: Vec<String>,
        /// The client's JWT.
        token: String,
        /// If resuming an existing sync session, the last LSN the client
        /// has already applied — the server resumes from just after this.
        resume_lsn: Option<PgLsn>,
        /// The protocol version this client speaks.
        protocol_version: u8,
    },
    /// Acknowledges receipt and application of all ops up to and including `lsn`.
    Ack {
        /// The highest LSN the client has applied.
        lsn: PgLsn,
    },
    /// A client-initiated write, to be applied upstream (e.g. to Postgres)
    /// and re-broadcast to other subscribers via the normal delta path.
    Write {
        /// The entity type being written.
        entity_type: String,
        /// The identifier of the entity being written.
        entity_id: String,
        /// The write operation itself.
        op: Op,
    },
    /// Liveness probe; the server should respond with a [`ServerMessage::Keepalive`].
    Ping,
}
